core/src/main/java/io/parsingdata/metal/data/DataExpressionSource.java

```diff
@@ -67,19 +66,20 @@ protected boolean isAvailable(final BigInteger offset, final BigInteger length)
 
     private synchronized byte[] getValue() {
         if (cache == null) {
-            final ImmutableList<Optional<Value>> results = dataExpression.eval(parseState, encoding);
+            final ImmutableList<Value> results = dataExpression.eval(parseState, encoding);
             if (results.size <= index) {
-                throw new IllegalStateException(format("ValueExpression dataExpression yields %d result(s) (expected at least %d).", results.size, index+1));
+                throw new IllegalStateException(format("ValueExpression dataExpression must yield at least %d results.", index+1));
             }
-            cache = getValueAtIndex(results, index, 0)
-                .computeResult()
-                .map(Value::getValue)
-                .orElseThrow(() -> new IllegalStateException(format("ValueExpression dataExpression yields empty Value at index %d.", index)));
+            final Value cacheValue = getValueAtIndex(results, index, 0).computeResult();
+            if (cacheValue == Value.NOT_A_VALUE) {
```

Comment:
You can `static import` `NOT_A_VALUE`.

Reply:
(Was already resolved when rewriting `==` checks to `equals(...)`.